File: TypeScript/Algorithm/Algorithm I/3. Longest Substring Without Repeating Characters.ts
/**
 * Returns the maximum of two numbers.
 * @param {number} valueA - The first number.
 * @param {number} valueB - The second number.
 * @returns {number} The larger of the two input numbers.
 */
function maxValue(valueA: number, valueB: number): number {
    return valueA > valueB ? valueA : valueB;
}

/**
 * Finds the length of the longest substring without repeating characters.
 * @param {string} s - The input string.
 * @returns {number} The length of the longest substring without repeating characters.
 */
function lengthOfLongestSubstring(s: string): number {
    const stringLength: number = s.length;
    let longestSubstring: number = 0;
    
    let startPointer: number = 0;

    const characterMap: Map<string, number> = new Map();

    for (let endPointer: number = 0; endPointer < stringLength; endPointer++) {
        const currentChar: string = s[endPointer];

        if (characterMap.has(currentChar) && characterMap.get(currentChar) >= startPointer)
            startPointer = characterMap.get(currentChar) + 1;
    
        longestSubstring = maxValue(longestSubstring, endPointer - startPointer + 1);

        characterMap.set(currentChar, endPointer);
    }

    return longestSubstring;
};

File: Java/Algorithm/Algorithm I/3. Longest Substring Without Repeating Characters.java
class Solution {

  /**
   * Calculates the length of the longest substring without repeating characters in a given string.
   *
   * @param s The input string.
   * @return The length of the longest substring without repeating characters.
   */
  public int lengthOfLongestSubstring(String s) {
    Map<Character, Integer> windowContext = new HashMap<>();
    int STRING_LENGTH = s.length();
    int longestSubstring = 0;
    int startPointer = 0;

    for (int endPointer = 0; endPointer < STRING_LENGTH; endPointer++) {
      char currentChar = s.charAt(endPointer);

      if (windowContext.containsKey(currentChar) == false || windowContext.get(currentChar) < startPointer) {
        windowContext.put(currentChar, endPointer);
        longestSubstring = Math.max(longestSubstring, endPointer - startPointer + 1);
      } 
      
      else {
        startPointer = windowContext.get(currentChar) + 1;
        windowContext.put(currentChar, endPointer);
      }
    }

    return longestSubstring;
  }
}
